protocol/consistencychecks.go

```diff
@@ -0,0 +1,265 @@
+	if str.Epoch > cc.CurrentEpoch {
+		cc.CurrentEpoch++
+		cc.SavedSTR = str.Signature
+	}
+	return
+}
+
+func (cc *ConsistencyChecks) verifyDirectoryProofs(dfs *DirectoryProofs,
+	uname string, key []byte) (int, error) {
+	// TODO: fix me
+	return invalidProof, nil
+}
+
+func (cc *ConsistencyChecks) verifyAuthPath(uname string, key []byte,
+	ap *m.AuthenticationPath,
+	str *m.SignedTreeRoot) (int, error) {
+	proofType := proofOfAbsence
+
+	// verify VRF Index
+	vrfKey := vrf.PublicKey(str.Policies.VrfPublicKey)
+	if !vrfKey.Verify([]byte(uname), ap.LookupIndex, ap.VrfProof) {
+		return invalidProof, ErrorBadVRFProof
+	}
+
+	if bytes.Equal(ap.LookupIndex, ap.Leaf.Index) { // proof of inclusion */
+		// verify name-to-key binding
+		// key is nil when the user does lookup for the first time
+		if key != nil && !ap.VerifyBinding(key) {
+			return invalidProof, ErrorBadBinding
+		}
+		// verify commitment
+		if !ap.Leaf.Commitment.Verify([]byte(uname), ap.Leaf.Value) {
+			return invalidProof, ErrorBadCommitment
+		}
+		proofType = proofOfInclusion
+	}
+
+	// verify auth path
+	if !ap.Verify(str.TreeHash) {
+		return invalidProof, ErrorBadAuthPath
+	}
+
+	return proofType, nil
+}
+
+func (cc *ConsistencyChecks) verifySTR(str *m.SignedTreeRoot) error {
+	// verify STR's signature
+	if !cc.signKey.Verify(str.Serialize(), str.Signature) {
+		return ErrorBadSignature
+	}
+
+	// verify hash chain
+	if (str.Epoch == cc.CurrentEpoch && bytes.Equal(cc.SavedSTR, str.Signature)) ||
+		(str.Epoch == cc.CurrentEpoch+1 && str.VerifyHashChain(cc.SavedSTR)) {
+		return nil
+	}
+	return ErrorBadSTR
+}
+
+// verifyReturnedPromise verifies the returned TB
+// included in the directory's response.
+func (cc *ConsistencyChecks) verifyReturnedPromise(tb *m.TemporaryBinding,
+	str *m.SignedTreeRoot,
+	ap *m.AuthenticationPath) error {
+	if cc.isUseTBs && tb != nil {
+		// verify TB's Signature
+		if !cc.signKey.Verify(tb.Serialize(str.Signature), tb.Signature) {
+			return ErrorBadSignature
+		}
+
+		// verify TB's VRF index
+		if !bytes.Equal(tb.Index, ap.LookupIndex) {
+			return ErrorBadIndex
+		}
+
+		cc.TBs = append(cc.TBs, tb)
+	}
+
+	return nil
+}
+
+// verifyFulfilledPromise verifies issued TBs was inserted
+// in the directory as promised.
+func (cc *ConsistencyChecks) verifyFulfilledPromise(str *m.SignedTreeRoot,
+	ap *m.AuthenticationPath) error {
+	if cc.isUseTBs {
+		backed := cc.TBs[:0]
+		for _, tb := range cc.TBs {
+			if cc.CurrentEpoch+1 == str.Epoch {
+				if !tb.Verify(ap) {
+					return ErrorBrokenPromise
+				}
+			} else if cc.CurrentEpoch == str.Epoch {
+				// keep current epoch's returned promises
+				// for future verifications
+				backed = append(backed, tb)
+			}
+		}
+
+		// clear all issued promises since they have been verified
+		// or the client has missed some epochs
+		cc.TBs = backed
+	}
+	return nil
+}
+
+func (cc *ConsistencyChecks) verifyProofTypeWithTB(statusCode ErrorCode, proofType int) ErrorCode {
```

Comment:
> the directory will return a ErrorNameExisted along with a proof of absence, if the client sends a registration request twice in the same epoch. This proof should be valid if the TB is enabled, but should be invalid otherwise

Oh ok, I understand what you're getting at now. I agree that the proof of absence should be valid in this case if TBs are enabled, but I'm not sure that the proof of absence is invalid otherwise. Does this mean we expect a proof of inclusion with TBs disabled? If we expect a proof of inclusion without the TB extension, that would mean that the server would have to publish a new STR immediately after each registration, wouldn't it? 

We haven't really talked about what the default server behavior is after registration. When we originally designed CONIKS, we didn't have TBs and we just assumed that clients would have to wait for the next epoch (1) to check that the registration succeeded (this has to happen anyway, with TBs there's proof) and (2) to actually start using the key for communication (TBs prevent this scenario). The server already returns a proof of absence after every registration by default, right?